openmls/src/framing/message_out.rs

```diff
@@ -0,0 +1,187 @@
+}
+
+/// MLSMessage (Body)
+///
+/// Note: Because [MlsMessageOutBody] already discriminates between
+/// `public_message`, `private_message`, etc., we don't use the
+/// `wire_format` field. This prevents inconsistent assignments
+/// where `wire_format` contradicts the variant given in `body`.
+///
+/// ```c
+/// // draft-ietf-mls-protocol-17
+/// struct {
+///     // ... continued from [MlsMessage] ...
+///
+///     WireFormat wire_format;
+///     select (MLSMessage.wire_format) {
+///         case mls_plaintext:
+///             PublicMessage plaintext;
+///         case mls_ciphertext:
+///             PrivateMessage ciphertext;
+///         case mls_welcome:
+///             Welcome welcome;
+///         case mls_group_info:
+///             GroupInfo group_info;
+///         case mls_key_package:
+///             KeyPackage key_package;
+///     }
+/// } MLSMessage;
+/// ```
+#[derive(Debug, PartialEq, Clone, TlsSerialize, TlsSize)]
+#[cfg_attr(feature = "test-utils", derive(TlsDeserialize))]
+#[repr(u8)]
+pub(crate) enum MlsMessageOutBody {
+    /// Plaintext message
+    #[tls_codec(discriminant = 1)]
+    PublicMessage(PublicMessage),
+
+    /// Ciphertext message
+    #[tls_codec(discriminant = 2)]
+    PrivateMessage(PrivateMessage),
+
+    /// Welcome message
+    #[tls_codec(discriminant = 3)]
+    Welcome(Welcome),
+
+    /// Group information
+    #[tls_codec(discriminant = 4)]
+    GroupInfo(GroupInfo),
+
+    /// KeyPackage
+    #[tls_codec(discriminant = 5)]
+    #[allow(dead_code)]
+    KeyPackage(KeyPackage),
+}
+
+impl From<PublicMessage> for MlsMessageOut {
+    fn from(public_message: PublicMessage) -> Self {
+        Self {
+            version: public_message.version(),
+            body: MlsMessageOutBody::PublicMessage(public_message),
+        }
+    }
+}
+
+impl From<GroupInfo> for MlsMessageOut {
+    fn from(group_info: GroupInfo) -> Self {
+        Self {
+            version: group_info.group_context().protocol_version(),
+            body: MlsMessageOutBody::GroupInfo(group_info),
+        }
+    }
+}
+
+impl MlsMessageOut {
+    /// Create an [`MlsMessageOut`] from a [`PrivateMessage`], as well as the
+    /// currently used [`ProtocolVersion`].
+    pub(crate) fn from_private_message(
+        private_message: PrivateMessage,
+        version: ProtocolVersion,
+    ) -> Self {
+        Self {
+            version,
+            body: MlsMessageOutBody::PrivateMessage(private_message),
+        }
+    }
+
+    /// Create an [`MlsMessageOut`] from a [`Welcome`] message and the currently
+    /// used [`ProtocolVersion`].
+    pub(crate) fn from_welcome(welcome: Welcome, version: ProtocolVersion) -> Self {
+        MlsMessageOut {
+            version,
+            body: MlsMessageOutBody::Welcome(welcome),
+        }
+    }
+
+    /// Turn an [`MlsMessageOut`] into a [`Welcome`].
+    #[cfg(any(feature = "test-utils", test))]
+    pub fn into_welcome(self) -> Option<Welcome> {
```

Comment:
Lets move the `#[cfg(any(feature = "test-utils", test))]` to a separate `impl` block to have nicer separation.